plugin/pkg/proxy/connect.go

```diff
@@ -117,6 +119,32 @@ func (p *Proxy) Connect(ctx context.Context, state request.Request, opts Options
 	for {
 		ret, err = pc.c.ReadMsg()
 		if err != nil {
+			if proto == "udp" {
+				// For UDP, if the error is an overflow, we probably have an upstream misbehaving in some way.
+				// (e.g. sending >512 byte responses without an eDNS0 OPT RR).
+				// Instead of returning an error, return an empty response with TC bit set. This will make the
+				// client retry over TCP (if that's supported) or at least receive a clean
+				// error. The connection is still good so we break before the close.
+				isDNSBufferError := false
+				isDNSOverflowError := false
+
+				// This is to handle a scenario in which upstream sets the TC bit, but doesn't truncate the response
+				// and we get ErrBuf instead of overflow.
+				if _, isDNSErr := err.(*dns.Error); isDNSErr && errors.Is(err, dns.ErrBuf) {
+					isDNSBufferError = true
+				} else if strings.Contains(err.Error(), "overflow") {
+					isDNSOverflowError = true
+				}
+
+				if isDNSBufferError || isDNSOverflowError {
```

Comment:
I think only one boolean is needed...
 
```suggestion
				shouldTruncate := false

				// This is to handle a scenario in which upstream sets the TC bit, but doesn't truncate the response
				// and we get ErrBuf instead of overflow.
				if _, isDNSErr := err.(*dns.Error); isDNSErr && errors.Is(err, dns.ErrBuf) {
					shouldTruncate = true
				} else if strings.Contains(err.Error(), "overflow") {
					shouldTruncate = true
				}

				if shouldTruncate {
```

Reply:
Ah yes yes. Thank you. I will do better next time.